File: c/2015/05/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif

char* subStr(char* str, char* dest, int16_t i) {
    dest[0] = str[i];
    dest[1] = str[i+1];
    dest[2] = '\0';
}

bool isNice1(char* str) {
    int a = 0;
    for (int16_t i = 0; i < strlen(str); i++) {
        if (str[i] == 'a' || str[i] == 'e' || str[i] == 'i' || str[i] == 'o' || str[i] == 'u') a++;
    }
    bool b = false;
    for (int16_t i = 0; i < strlen(str) - 1; i++) {
        b |= str[i] == str[i+1];
    }
    bool c = true;
    for (int16_t i = 0; i < strlen(str) - 1; i++) {
        char tmp[3] = "";
        subStr(str, tmp, i);
        c &= strcmp(tmp, "ab") && strcmp(tmp, "cd") && strcmp(tmp, "pq") && strcmp(tmp, "xy");
    }
    return a >= 3 && b && c;
}

bool isNice2(char* str) {
    bool a = false;
    for (int16_t i = 0; i < strlen(str) - 1; i++) {
        char tmp[3] = "";
        subStr(str, tmp, i);
        for (int16_t j = i+2; j < strlen(str) - 1; j++) {
            char tmp2[3] = "";
            subStr(str, tmp2, j);
            a |= !strcmp(tmp, tmp2); 
        }
    }
    bool b = false;
    for (int16_t i = 0; i < strlen(str) - 2; i++) {
        b |= str[i] == str[i+2];
    }
    return a && b;
}

void part1() {
    ARR_STRING* in = readInput(FILE_NAME);
    int64_t k = 0;
    for (int16_t i = 0; i < in->length; i++) {
        if (isNice1(in->strings[i])) k++;
    }
    printf("Part 1: %ld\n", k);
    freeArrString(in);
}

void part2() {
    ARR_STRING* in = readInput(FILE_NAME);
    int64_t k = 0;
    for (int16_t i = 0; i < in->length; i++) {
        if (isNice2(in->strings[i])) k++;
    }
    printf("Part 2: %ld\n", k);
    freeArrString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}

File: python/2015/05/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def is_nice1(str):
    a = sum([1 if a in "aeiou" else 0 for a in str]) >= 3 
    b = sum([1 if str[i] == str[i+1] else 0 for i in range(len(str)-1)]) > 0
    c = sum([1 if str[i:i+2] in ["ab", "cd", "pq", "xy"] else 0 for i in range(len(str)-1)]) == 0
    return a and b and c

def is_nice2(str):
    a = sum([1 if str[i:i+2] in str[i+2:] else 0 for i in range(len(str)-2)]) > 0
    b = sum([1 if str[i] == str[i+2] else 0 for i in range(len(str)-2)]) > 0
    return a and b

def read_input():
    return u.read_lines(file)

def part1():
    A = read_input()
    k = sum([1 if is_nice1(a) else 0 for a in A])
    print("Part 1:",k)

def part2():
    A = read_input()
    k = sum([1 if is_nice2(a) else 0 for a in A])
    print("Part 2:",k)

part1()
part2()
